File: frontend/src/sdk/messages.ts
import type { ProductId, VariantId } from "./events";

export type Viewport = {
  width: number;
  height: number;
  device_scale_factor: number;
};

export type DomActionSummary = {
  action: string;
  label: string;
  selector: string;
  product_id: ProductId | null;
  variant_id: VariantId | null;
  requires_confirmation: boolean;
};

export type BrowserDomSummary = {
  visible_agent_actions: DomActionSummary[];
  selected_variant_id: VariantId | null;
  cart_count: number;
  cart_product_ids: ProductId[];
};

export type BrowserObservation = {
  url: string;
  screenshot: string;
  viewport: Viewport;
  dom_summary: BrowserDomSummary;
};

export type MerchantEventCreate = {
  type:
    | "guided_session_opened"
    | "simulation_opened"
    | "simulation_started"
    | "assistant_opened"
    | "guide_allowed"
    | "product_viewed"
    | "variant_selected"
    | "add_to_cart_clicked"
    | "cart_updated"
    | "loyalty_prompt_seen"
    | "loyalty_accepted";
  source: "merchant_sdk" | "merchant_harness" | "external_agent";
  product_id?: ProductId | null;
  variant_id?: VariantId | null;
  message?: string | null;
};

export type ComputerAction =
  | { type: "click"; action_id: string; x: number; y: number; reason: string }
  | { type: "double_click"; action_id: string; x: number; y: number; reason: string }
  | { type: "scroll"; action_id: string; x?: number | null; y?: number | null; scroll_x?: number | null; scroll_y?: number | null; reason: string }
  | { type: "type"; action_id: string; text: string; reason: string }
  | { type: "wait"; action_id: string; reason: string }
  | { type: "keypress"; action_id: string; key: string; reason: string }
  | { type: "move"; action_id: string; x: number; y: number; reason: string }
  | { type: "drag"; action_id: string; x: number; y: number; reason: string }
  | { type: "screenshot"; action_id: string; reason: string };

export type ActionExpectation = {
  expected_state_change: string;
  product_id: ProductId | null;
  variant_id: VariantId | null;
};

export type BrowserActionResult = {
  action_id: string;
  success: boolean;
  url: string;
  observation: BrowserObservation;
  events: MerchantEventCreate[];
  message: string | null;
};

export type GuideReadyMessage = {
  type: "guide_ready";
  session_id: string;
  observation: BrowserObservation;
};

export type ObservationMessage = {
  type: "observation";
  session_id: string;
  observation: BrowserObservation;
};

export type ActionResultMessage = {
  type: "action_result";
  session_id: string;
  result: BrowserActionResult;
};

export type BrowserToBackendMessage = GuideReadyMessage | ObservationMessage | ActionResultMessage;

export type RequestObservationMessage = {
  type: "request_observation";
  session_id: string;
};

export type ExecuteActionMessage = {
  type: "execute_action";
  session_id: string;
  action_id: string;
  action: ComputerAction;
  expected: ActionExpectation;
};

export type AssistantUpdateMessage = {
  type: "assistant_update";
  session_id: string;
  message: string;
};

export type TraceUpdateMessage = {
  type: "trace_update";
  session_id: string;
  trace: {
    trace_id: string;
    verification: { status: string; message: string };
    action?: ComputerAction | null;
  };
};

export type GuideDoneMessage = {
  type: "guide_done";
  session_id: string;
  message: string;
};

export type GuideErrorMessage = {
  type: "guide_error";
  session_id: string;
  message: string;
};

export type ConfirmationRequestMessage = {
  type: "confirmation_request";
  session_id: string;
  message: string;
};

export type BackendToBrowserMessage =
  | RequestObservationMessage
  | ExecuteActionMessage
  | AssistantUpdateMessage
  | TraceUpdateMessage
  | GuideDoneMessage
  | GuideErrorMessage
  | ConfirmationRequestMessage;

export function parseBackendMessage(payload: string): BackendToBrowserMessage {
  const parsed = JSON.parse(payload) as BackendToBrowserMessage;
  return parsed;
}

File: backend/app/realtime/messages.py
from __future__ import annotations

from typing import Annotated, Literal

from pydantic import BaseModel, Field, TypeAdapter

from app.ids import ActionId, SessionId
from app.models import (
    ActionExpectation,
    BrowserActionResult,
    BrowserObservation,
    ComputerAction,
    TraceEntry,
)


class GuideReadyMessage(BaseModel):
    type: Literal["guide_ready"]
    session_id: SessionId
    observation: BrowserObservation


class ObservationMessage(BaseModel):
    type: Literal["observation"]
    session_id: SessionId
    observation: BrowserObservation


class ActionResultMessage(BaseModel):
    type: Literal["action_result"]
    session_id: SessionId
    result: BrowserActionResult


class CustomerMessage(BaseModel):
    type: Literal["customer_message"]
    session_id: SessionId
    message: str


class CustomerApprovalMessage(BaseModel):
    type: Literal["customer_approval"]
    session_id: SessionId
    approved: bool


BrowserToBackendMessage = Annotated[
    GuideReadyMessage
    | ObservationMessage
    | ActionResultMessage
    | CustomerMessage
    | CustomerApprovalMessage,
    Field(discriminator="type"),
]


class RequestObservationMessage(BaseModel):
    type: Literal["request_observation"] = "request_observation"
    session_id: SessionId


class ExecuteActionMessage(BaseModel):
    type: Literal["execute_action"] = "execute_action"
    session_id: SessionId
    action_id: ActionId
    action: ComputerAction
    expected: ActionExpectation


class AssistantUpdateMessage(BaseModel):
    type: Literal["assistant_update"] = "assistant_update"
    session_id: SessionId
    message: str


class ConfirmationRequestMessage(BaseModel):
    type: Literal["confirmation_request"] = "confirmation_request"
    session_id: SessionId
    message: str


class TraceUpdateMessage(BaseModel):
    type: Literal["trace_update"] = "trace_update"
    session_id: SessionId
    trace: TraceEntry


class GuideDoneMessage(BaseModel):
    type: Literal["guide_done"] = "guide_done"
    session_id: SessionId
    message: str


class GuideErrorMessage(BaseModel):
    type: Literal["guide_error"] = "guide_error"
    session_id: SessionId
    message: str


BackendToBrowserMessage = (
    RequestObservationMessage
    | ExecuteActionMessage
    | AssistantUpdateMessage
    | ConfirmationRequestMessage
    | TraceUpdateMessage
    | GuideDoneMessage
    | GuideErrorMessage
)


browser_message_adapter: TypeAdapter[BrowserToBackendMessage] = TypeAdapter(BrowserToBackendMessage)


def parse_browser_message(payload: str) -> BrowserToBackendMessage:
    return browser_message_adapter.validate_json(payload)

